### PR TITLE
show README date

### PR DESCRIPTION
**Pull request by eriktks**
Refs #57

## Background

It is possible that howfairis checks an old version of a recently updated github README. To make detection of this behavior easier, a warning message should be shown if the README was updated less than 5 minutes ago 

## Describe the changes made in this pull request

`howfairis/workarounds/github_caching.py`:
1. get list of commits of README that happened in the recent 5 minutes
2. if the list is non-empty: print warning message

`howfairis/cli/print_call_to_action.py`:
1. only call `github_caching.py` check for Github repos (problem only exists on Github)
2. only call `github_caching.py` check when quiet option is not set

`tests/test_github_caching.py`
1. new tests for `howfairis/workarounds/github_caching.py`

`tests/test_cli.py`
1. new tests for `howfairis/cli/print_call_to_action.py`
 
## Instructions to review the pull request

1. Rebuild howfairis
2. Check any own github repository `howfairis https://github.com/me/repository` to verify that howfairis does not generate warning
3. Update README of repository and re-check within 5 minutes to see the warning message at the end of the report
4. an alternative for point 3: run: `pytest tests/test_github_caching.py` (this configuraion is tested in the function `test_github_readme_update_check_critical_time`)

### CODE REVIEW

**Comment by eriktks** [comment]
Jurriaan suggested two changes:

1. only show the README creation date/time if it was less than 5 minutes ago
2. only show the date/time if the badge is different from the current badge

The first suggestion is fine. The second one will probably not help the user because we want to warn the user about a potential problem when the predicted badge is different from what the user expects but we do not know what the user expects. So it seems better to always warn the user when situation 1. arises.

File: howfairis/cli.py [hunk 1]
```diff
@@ -94,26 +100,53 @@ def cli(url=None, branch=None, config_file=None, remote_config_file=None, path=N
...
     print("The compliance of your repository is different from the current " +
           "fair-software.eu badge in your " +
           "{0}. Please replace it with the following snippet:\n\n{1}"
           .format(checker.readme.filename, badge))
-
+    github_readme_creation_check(url, checker.readme.filename, checker.repo.platform, branch)
     sys.exit(1)
 
 
+def github_readme_creation_check(url, filename, platform, branch):
```
Comment (jspaaks) [inline howfairis/cli.py:127]: This is probably not the best place to put this functionality

File: howfairis/cli.py [hunk 2]
```diff
@@ -94,26 +100,53 @@ def cli(url=None, branch=None, config_file=None, remote_config_file=None, path=N
...
+    if platform != Platform.GITHUB:
+        return(0)
```
Comment (jspaaks) [inline howfairis/cli.py:129]: Do you intend to use the int 0? If you do, `return 0` is the correct style

Comment (jspaaks) [reply howfairis/cli.py:129]: Stefan pointed out that we should probably raise or warn depending on whether we have already implemented something. E.g. `raise NotImplementedException`

File: howfairis/cli.py [hunk 3]
```diff
@@ -94,26 +100,53 @@ def cli(url=None, branch=None, config_file=None, remote_config_file=None, path=N
...
+    if branch is None:
+        branch = "master"
```
Comment (jspaaks) [inline howfairis/cli.py:131]: `master` is probably not the best value here, check the code for `Repo`

File: howfairis/cli.py [hunk 4]
```diff
@@ -94,26 +100,53 @@ def cli(url=None, branch=None, config_file=None, remote_config_file=None, path=N
...
+    try:
+        response = requests.get(url+"/blob/"+branch+"/"+filename)
```
Comment (jspaaks) [inline howfairis/cli.py:133]: syntax linters ask for space between operators

File: howfairis/cli.py [hunk 5]
```diff
@@ -94,26 +100,53 @@ def cli(url=None, branch=None, config_file=None, remote_config_file=None, path=N
...
+        date_string = BeautifulSoup(response.text, "html.parser").select("relative-time")[0]["datetime"]
```
Comment (jspaaks) [inline howfairis/cli.py:134]: I'd prefer to use the API instead of BeautifulSoup

**Review by jspaaks — CHANGES_REQUESTED** [review]

File: howfairis/cli.py (hunk 1 above)
Comment (eriktks) [reply howfairis/cli.py:131]: Ok, the function was moved from cli.py to checker.py

File: howfairis/cli.py (hunk 5 above)
Comment (eriktks) [reply howfairis/cli.py:138]: Ok, the call to BeautifulSoup has been replaced with a call to the Github API

File: howfairis/cli.py (hunk 3 above)
Comment (eriktks) [reply howfairis/cli.py:135]: Ok. The branch variable was not required anymore after switching from BeautifulSoup to the Github API

File: howfairis/cli.py (hunk 2 above)
Comment (eriktks) [reply howfairis/cli.py:133]: Ok, the numbers have been removed from the return statements. Also this if statement has been moved up one layer to make clear that the check should only be performed for Github repos

File: howfairis/cli.py (hunk 4 above)
Comment (eriktks) [reply howfairis/cli.py:137]: Ok, the string additions have been replaced by f"{...}"

**Comment by jspaaks** [comment]
Hi Erik, I see you're also getting the API docs error I had. I solved it with this https://github.com/fair-software/howfairis/pull/208#issuecomment-779361314.

File: howfairis/workarounds/github_caching.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+from datetime import datetime
+from datetime import timedelta
+import requests
+from dateutil import tz
+
+
+def github_caching_check(checker):
+    try:
+        response = requests.get(checker.repo.api)
+        date_created_string = response.json().get("created_at")
```
Comment (jspaaks) [inline howfairis/workarounds/github_caching.py:10]: Hi Erik, can you point me to where the meaning of this key is described?

Comment (eriktks) [reply howfairis/workarounds/github_caching.py:10]: Sorry, I do not understand the word "key" in this context. Do you mean "created at"? I got that from an example output the Github API generated.

File: .gitignore [hunk 7]
```diff
@@ -9,3 +9,8 @@
 
 docs/_build
 docs/apidocs
+Makefile
```
Comment (fdiblen) [inline .gitignore:12]:
```suggestion
```

File: .gitignore [hunk 8]
```diff
@@ -9,3 +9,8 @@
...
+000README
```
Comment (fdiblen) [inline .gitignore:13]:
```suggestion
```

File: .gitignore [hunk 9]
```diff
@@ -9,3 +9,8 @@
...
+tmp/
```
Comment (fdiblen) [inline .gitignore:14]:
```suggestion
```

File: .gitignore [hunk 10]
```diff
@@ -9,3 +9,8 @@
...
+howfairis/OldFiles
```
Comment (fdiblen) [inline .gitignore:15]:
```suggestion
```

File: .gitignore [hunk 11]
```diff
@@ -9,3 +9,8 @@
...
+tests/OldFiles
```
Comment (fdiblen) [inline .gitignore:16]:
```suggestion
```

File: howfairis/cli/print_call_to_action.py [hunk 12]
```diff
@@ -1,34 +1,37 @@
...
+                  .format(checker.readme.filename, badge))
 
-    if not is_quiet:
+    elif not is_quiet:
         print("The compliance of your repository is different from the current " +
               "fair-software.eu badge in your " +
               "{0}. Please replace it with the following snippet:\n\n{1}"
-              .format(readme.filename, badge))
+              .format(checker.readme.filename, badge))
+        github_caching_check(checker)
```
Comment (fdiblen) [inline howfairis/cli/print_call_to_action.py:35]: Do we need this function call here?

File: howfairis/workarounds/github_caching.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
+from datetime import datetime
+from datetime import timedelta
+import requests
+from dateutil import tz
+
+
+def github_caching_check(checker):
```
Comment (fdiblen) [inline howfairis/workarounds/github_caching.py:7]: We can use GitHub API to get info about the commits. https://docs.github.com/en/rest/reference/repos#commits

we can use `path` parameter to point to README.md or README.rst
and then get timestamp info

**Review by fdiblen — CHANGES_REQUESTED** [review]
great work! suggested some changes. :sake:

File: howfairis/workarounds/github_caching.py (hunk 13 above)
Comment (jspaaks) [reply howfairis/workarounds/github_caching.py:7]: In addition to what Faruk suggests, I also tested with `?page=0&per_page=1`, so you get an array of 1 element back

File: howfairis/cli/print_call_to_action.py (hunk 12 above)
Comment (eriktks) [reply howfairis/cli/print_call_to_action.py:37]: Ok, the function call was moved to the right place

File: howfairis/workarounds/github_caching.py (hunk 13 above)
Comment (eriktks) [reply howfairis/workarounds/github_caching.py:7]: Ok, both suggestions have been implemented in `howfairis/workarounds/github_caching.py`

**Comment by fdiblen** [comment]
I see we are back to draft again. @eriktks is this ready to review?

File: howfairis/workarounds/github_caching.py [hunk 14]
```diff
@@ -1,2 +1,19 @@
...
+    try:
+        date_critical_utc = datetime.now().replace(second=0).astimezone(tz.tzutc()) - timedelta(minutes=5)
+        date_critical_utc_string = date_critical_utc.strftime("%Y-%m-%dT%H:%M:%SZ")
```
Comment (jspaaks) [inline howfairis/workarounds/github_caching.py:10]: This seems a bit complicated to me and I think it introduces another dependency right?

Does this suffice perhaps:

```python
crit = datetime.now(timezone.utc) - timedelta(minutes=5)
```

and then in the request use `crit.isoformat()`.

While you're at it, maybe change to use the formatted string style that's used throughout the rest of the library

Comment (eriktks) [reply howfairis/workarounds/github_caching.py]: This is a good idea. Implemented!

**Comment by jspaaks** [comment]
Erik can you check the dependencies, my editor is warning me about an unsatisfied requirement

**Comment by jspaaks** [comment]
Also see my earlier comment about how this project does formatted strings

File: tests/test_github_caching.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    filename = "README.rst"
+    url = f"https://github.com/{owner}/{repo_string}"
+    requests_mock.get(f"https://raw.githubusercontent.com/{owner}/{repo_string}/main/{filename}",
+                      json={}, status_code=200)
+    requests_mock.get(f"https://raw.githubusercontent.com/{owner}/{repo_string}/main/.howfairis.yml",
+                      json={}, status_code=200)
+    repo = Repo(url, branch="main")
+    checker = Checker(repo)
+    capsys.readouterr()
+    return(checker)
```
Comment (jspaaks) [inline tests/test_github_caching.py:19]: parentheses

File: tests/test_github_caching.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
+from requests_mock import Mocker
+from howfairis.checker import Checker
+from howfairis.repo import Repo
+from howfairis.workarounds.github_caching import github_caching_check
+
+
+def initialize(requests_mock: Mocker, capsys):
```
Comment (jspaaks) [inline tests/test_github_caching.py:7]:
```suggestion
def initialize_checker(requests_mock: Mocker, capsys):
```

File: tests/test_github_caching.py [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+                      json={}, status_code=200)
+    requests_mock.get(f"https://raw.githubusercontent.com/{owner}/{repo_string}/main/.howfairis.yml",
+                      json={}, status_code=200)
+    repo = Repo(url, branch="main")
+    checker = Checker(repo)
+    capsys.readouterr()
+    return(checker)
+
+
+def test_github_readme_update_check_critical_time(requests_mock: Mocker, capsys):
```
Comment (jspaaks) [inline tests/test_github_caching.py:22]:
```suggestion
def test_github_caching _should_warn(requests_mock: Mocker, capsys):
```

File: tests/test_github_caching.py [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    checker = initialize(requests_mock, capsys)
+    requests_mock.get(f"{checker.repo.api}/commits", json=[0], status_code=200)
```
Comment (jspaaks) [inline tests/test_github_caching.py:24]:
```suggestion
    requests_mock.get(f"{checker.repo.api}/commits", json=["nonzero array length should trigger warning"], status_code=200)
```

File: tests/test_github_caching.py [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    github_caching_check(checker)
+    actual_out_err = capsys.readouterr()
+    expected_out = (f"Warning: Your {checker.readme.filename} was updated less than 5 minutes ago. " +
+                    "The effects of this update are not visible yet in the calculated compliance.\n")
+    assert actual_out_err[0] == expected_out
+
+
+def test_github_readme_update_check_fine_time(requests_mock: Mocker, capsys):
```
Comment (jspaaks) [inline tests/test_github_caching.py:32]:
```suggestion
def test_github_caching_should_not_warn(requests_mock: Mocker, capsys):
```

File: tests/test_github_caching.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+    checker = initialize(requests_mock, capsys)
+    requests_mock.get(f"{checker.repo.api}/commits", json=[], status_code=200)
+    github_caching_check(checker)
+    actual_out_err = capsys.readouterr()
+    expected_out = ""
+    assert actual_out_err[0] == expected_out
+
+
+def test_github_readme_update_check_no_readme(requests_mock: Mocker, capsys):
```
Comment (jspaaks) [inline tests/test_github_caching.py:41]:
```suggestion
def test_github_caching_without_readme(requests_mock: Mocker, capsys):
```

**Review by jspaaks — CHANGES_REQUESTED** [review]
Hi Erik, thanks for making the explanation at the top so elaborate, it helped me. I left a couple of comments. You can use the github interface to commit these if you agree with them.

**Comment by jspaaks** [comment]
If I remember correctly, we said we werent gonna test functions that were not part of the puiblic interface. You're testing `print_call_to_action.py`. This is something you can point out to the rest of the team during standup, to verify you're still working in the same direction as the others.

File: tests/test_github_caching.py (hunk 15 above)
Comment (eriktks) [reply tests/test_github_caching.py:19]: Ok

**Comment by eriktks** [comment]
> If I remember correctly, we said we weren't gonna test functions that were not part of the public interface. You're testing `print_call_to_action.py`. This is something you can point out to the rest of the team during standup, to verify you're still working in the same direction as the others.

This code was written when all of cli.py was in  single file. Since then the error messages have been moved to separate files, not part of the public interface. The tests related to these messages have been removed from tests/test_cli.py. The remaining tests involve checks of the exit code. Perhaps they are now redundant because of other tests that were added recently.

File: tests/test_github_caching.py (hunk 18 above)
Comment (eriktks) [reply tests/test_github_caching.py]: This suggestion was committed via the command line

**Comment by eriktks** [comment]
> Also see my earlier comment about how this project does formatted strings

Yes, this has been taken care

**Comment by eriktks** [comment]
> Erik can you check the dependencies, my editor is warning me about an unsatisfied requirement

Yes, done

File: howfairis/repo.py [hunk 21]
```diff
@@ -116,5 +116,5 @@ def _get_default_branch(self):
         try:
             response.raise_for_status()
         except requests.HTTPError as e:
-            raise GetDefaultBranchException("Something went wrong asking the repo for its default branch.") from e
+            raise GetDefaultBranchException("Something went wrong asking the repo for its default branch: " + str(e)) from e
```
Comment (jspaaks) [inline howfairis/repo.py:119]: Why did you add the `str(e)` here?

File: tests/case0/github/test_repo_no_args.py [hunk 22]
```diff
@@ -22,7 +23,7 @@ def test_branch(self, mocker: Mocker):
     def test_default_branch(self, mocker: Mocker):
         with mocker, pytest.raises(GetDefaultBranchException) as exc_info:
             get_repo()
-        assert str(exc_info.value) == "Something went wrong asking the repo for its default branch."
+        assert re.search("Something went wrong asking the repo for its default branch.", str(exc_info.value))
```
[the remark this answers is not in the document]
Comment (jspaaks) [reply tests/case0/github/test_repo_no_args.py:26]: or `str.startswith(substr)`

**Review by jspaaks — CHANGES_REQUESTED** [review]
Left a couple more comments

File: tests/test_cli.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
+import re
```
[the remark this answers is not in the document]
Comment (eriktks) [reply tests/test_cli.py:1]: Correct, I have now removed it

File: howfairis/workarounds/github_caching.py [hunk 24]
```diff
@@ -1,2 +1,22 @@
...
+                                    "per_page": 1,
+                                    "path": checker.readme.filename,
+                                    "since": critical_time.strftime("%Y-%m-%dT%H:%M:%SZ")
+        })
+        if len(response.json()) > 0:
+            print(("Warning: Your {0} was updated less than 5 minutes ago. The effects of this update are not " +
+                   "visible yet in the calculated compliance.").format(checker.readme.filename))
+        return
+    except ValueError:
+        return
```
[the remark this answers is not in the document]
Comment (eriktks) [reply howfairis/workarounds/github_caching.py:22]: I could not find a fitting error so I went for a general error. Now I have replaced it with `requests.exceptions.ConnectionError` which makes sense and makes the linter happy

File: tests/case0/github/test_repo_no_args.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (eriktks) [reply tests/case0/github/test_repo_no_args.py:26]: Yes, the error message in `howfairis/repo.py` was extended so string comparison did not work anymore. I have now used `str.startswith` in both this file and the equivalent one under gitlab

File: howfairis/repo.py (hunk 21 above)
Comment (eriktks) [reply howfairis/repo.py:119]: Because the error message was not clear enough. During testing this error occurred and only after including str(e) here, I could find out what was wrong. I hope this addition will save other testers some time.

File: tests/test_cli.py [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+from click.testing import CliRunner
+from requests_mock import Mocker
+from howfairis.cli.cli import cli
+
+
+runner = CliRunner()
```
[the remark this answers is not in the document]
Comment (eriktks) [reply tests/test_cli.py:7]: No, this is indeed redundant. I have removed the line.

File: tests/test_cli.py [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+    requests_mock.get(url, status_code=200)
+    requests_mock.get(api, json={"default_branch": "master"}, status_code=200)
+    requests_mock.get(api + "/license", status_code=200)
+    requests_mock.get(raw + "/.howfairis.yml", status_code=200)
+    requests_mock.get(raw + "/CITATION", status_code=200)
+    requests_mock.get(raw + "/CITATION.cff", status_code=200)
+    requests_mock.get(raw + "/codemeta.json", status_code=200)
+    requests_mock.get(raw + "/" + filename, text=howfairis_badge+pypi_badge+cii_badge, status_code=200)
+    requests_mock.get(raw + "/.zenodo.json", status_code=200)
+    requests_mock.get(api + "/commits", status_code=200)
```
[the remark this answers is not in the document]
Comment (eriktks) [reply tests/test_cli.py]: Ok, this update was added in a local editor (four times)

File: howfairis/repo.py (hunk 21 above)
Comment (jspaaks) [reply howfairis/repo.py:119]: Ok, can you explain what `from e` means in the snippet above?

File: howfairis/workarounds/github_caching.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (jspaaks) [reply howfairis/workarounds/github_caching.py]: Did you try triggering an error an see what Exceptions are actually raised, for example by using a debugger?

File: tests/test_cli.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (jspaaks) [reply tests/test_cli.py]: Did your editor not warn you? We should take a look at your setup --it's quite inefficient having this kind of back and forth

File: howfairis/workarounds/github_caching.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (jspaaks) [reply howfairis/workarounds/github_caching.py]: OK, in the interest of merging this quickly, replace `ValueError` with `Exception`. SonarCloud will complain about raising a general exception, we can deal with it then.

File: tests/case0/github/test_repo_no_args.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (jspaaks) [reply tests/case0/github/test_repo_no_args.py]: Revert your changes, use `from e` (see below)

**Review by jspaaks — CHANGES_REQUESTED** [review]
Hopefully the last couple changes ...

**Review by fdiblen — APPROVED** [review]
Gooood job! :medal_military:

**Review by jspaaks — APPROVED** [review]
marathon man!

**Comment by eriktks** [comment]
Thank you for your patience!
